File: packages/php/src/Storage/Postgres/PostgresStorageAdapter.php
<?php

declare(strict_types=1);

namespace Greyfoundry\Identifold\Storage\Postgres;

use Greyfoundry\Identifold\IdentifoldException;
use Greyfoundry\Identifold\Storage\ReferenceLookup;
use Greyfoundry\Identifold\Storage\ReferenceMapping;
use Greyfoundry\Identifold\Storage\ReferenceReservation;
use Greyfoundry\Identifold\Storage\ReferenceStore;
use Greyfoundry\Identifold\Storage\SequenceAllocationRequest;
use Greyfoundry\Identifold\Storage\SequenceAllocator;
use PDO;
use PDOException;

final readonly class PostgresStorageAdapter implements ReferenceStore, ReferenceLookup, SequenceAllocator
{
    public function __construct(private PDO $connection)
    {
    }

    public function reserve(ReferenceReservation $request): bool
    {
        $value = $this->scalar(
            'SELECT identifold_reserve_reference(:machine_id::text::uuid, :namespace::text, :reference::text)',
            [
                'machine_id' => $request->machineId,
                'namespace' => $request->namespace,
                'reference' => $request->reference,
            ],
        );
        if ($value === true || $value === 't' || $value === 1 || $value === '1') {
            return true;
        }
        if ($value === false || $value === 'f' || $value === 0 || $value === '0') {
            return false;
        }
        throw new IdentifoldException('allocation_conflict');
    }

    public function resolve(string $reference, string $namespace): ?ReferenceMapping
    {
        try {
            $statement = $this->connection->prepare(
                'SELECT resolved_machine_id::text AS machine_id, resolved_namespace AS namespace '
                . 'FROM identifold_resolve_reference(:reference::text, :namespace::text)',
            );
            $statement->execute(['reference' => $reference, 'namespace' => $namespace]);
            $row = $statement->fetch(PDO::FETCH_ASSOC);
            if ($row === false) {
                return null;
            }
            if (!is_string($row['machine_id'] ?? null) || !is_string($row['namespace'] ?? null)) {
                throw new IdentifoldException('allocation_conflict');
            }
            return new ReferenceMapping($row['machine_id'], $row['namespace']);
        } catch (PDOException $exception) {
            throw $this->map($exception);
        }
    }

    public function allocate(SequenceAllocationRequest $request): int
    {
        $value = $this->scalar(
            'SELECT identifold_allocate_sequence('
            . ':machine_id::text::uuid, :namespace::text, :prefix::text, :scope::text, :width::smallint)',
            [
                'machine_id' => $request->machineId,
                'namespace' => $request->namespace,
                'prefix' => $request->referencePrefix,
                'scope' => $request->scope,
                'width' => $request->width,
            ],
        );
        if ((is_int($value) || (is_string($value) && preg_match('/^\d+$/D', $value) === 1))
            && (int) $value >= 0) {
            return (int) $value;
        }
        throw new IdentifoldException('allocation_conflict');
    }

    /** @param array<string, int|string|null> $parameters */
    private function scalar(string $sql, array $parameters): mixed
    {
        try {
            $statement = $this->connection->prepare($sql);
            $statement->execute($parameters);
            return $statement->fetchColumn();
        } catch (PDOException $exception) {
            throw $this->map($exception);
        }
    }

    private function map(PDOException $exception): IdentifoldException
    {
        $state = $exception->errorInfo[0] ?? $exception->getCode();
        return new IdentifoldException(match ($state) {
            '22003' => 'sequence_overflow',
            '22023' => 'invalid_allocation_policy',
            default => 'allocation_conflict',
        });
    }
}

File: packages/java/src/main/java/io/greyfoundry/identifold/storage/postgres/PostgresStorageAdapter.java
package io.greyfoundry.identifold.storage.postgres;

import io.greyfoundry.identifold.Identifold;
import io.greyfoundry.identifold.storage.ReferenceMapping;
import io.greyfoundry.identifold.storage.ReferenceReservation;
import io.greyfoundry.identifold.storage.SequenceAllocationRequest;
import io.greyfoundry.identifold.storage.StorageAdapter;
import java.sql.SQLException;
import java.sql.Types;
import java.util.Objects;
import java.util.concurrent.CompletionException;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.Executor;
import java.util.concurrent.CompletableFuture;
import javax.sql.DataSource;

public final class PostgresStorageAdapter implements StorageAdapter {
    private final DataSource dataSource;
    private final Executor executor;

    public PostgresStorageAdapter(DataSource dataSource, Executor executor) {
        this.dataSource = Objects.requireNonNull(dataSource, "dataSource");
        this.executor = Objects.requireNonNull(executor, "executor");
    }

    @Override
    public CompletionStage<Boolean> reserve(ReferenceReservation request) {
        return CompletableFuture.supplyAsync(() -> {
            try (var connection = dataSource.getConnection();
                 var statement = connection.prepareStatement(
                     "SELECT identifold_reserve_reference(?::text::uuid, ?::text, ?::text)")) {
                statement.setString(1, request.machineId());
                statement.setString(2, request.namespace());
                statement.setString(3, request.reference());
                try (var result = statement.executeQuery()) {
                    if (!result.next()) throw failure("allocation_conflict");
                    return result.getBoolean(1);
                }
            } catch (SQLException exception) {
                throw failure(mapCode(exception));
            }
        }, executor);
    }

    @Override
    public CompletionStage<ReferenceMapping> resolve(String reference, String namespace) {
        return CompletableFuture.supplyAsync(() -> {
            try (var connection = dataSource.getConnection();
                 var statement = connection.prepareStatement(
                     "SELECT resolved_machine_id::text, resolved_namespace "
                         + "FROM identifold_resolve_reference(?::text, ?::text)")) {
                statement.setString(1, reference);
                statement.setString(2, namespace);
                try (var result = statement.executeQuery()) {
                    if (!result.next()) return null;
                    return new ReferenceMapping(result.getString(1), result.getString(2));
                }
            } catch (SQLException exception) {
                throw failure(mapCode(exception));
            }
        }, executor);
    }

    @Override
    public CompletionStage<Long> allocate(SequenceAllocationRequest request) {
        return CompletableFuture.supplyAsync(() -> {
            try (var connection = dataSource.getConnection();
                 var statement = connection.prepareStatement(
                     "SELECT identifold_allocate_sequence(?::text::uuid, ?::text, ?::text, ?::text, ?::smallint)")) {
                statement.setString(1, request.machineId());
                statement.setString(2, request.namespace());
                statement.setString(3, request.referencePrefix());
                if (request.scope() == null) statement.setNull(4, Types.VARCHAR);
                else statement.setString(4, request.scope());
                statement.setInt(5, request.width());
                try (var result = statement.executeQuery()) {
                    if (!result.next()) throw failure("allocation_conflict");
                    var sequence = result.getLong(1);
                    if (result.wasNull() || sequence < 0) throw failure("allocation_conflict");
                    return sequence;
                }
            } catch (SQLException exception) {
                throw failure(mapCode(exception));
            }
        }, executor);
    }

    private static String mapCode(SQLException exception) {
        if ("22003".equals(exception.getSQLState())) return "sequence_overflow";
        if ("22023".equals(exception.getSQLState())) return "invalid_allocation_policy";
        return "allocation_conflict";
    }

    private static CompletionException failure(String code) {
        return new CompletionException(new Identifold.IdentifoldException(code));
    }
}
